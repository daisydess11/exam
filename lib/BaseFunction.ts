import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";
import { Duration } from "aws-cdk-lib";

interface BaseFunctionProps {
    TABLE_NAME: string;
    SECONDARY_INDEX_NAME?: string;
    TOPIC_ARN?: string;
}

export class BaseFunction extends NodejsFunction {
    constructor(scope: Construct, id:string,props:BaseFunctionProps){
        super(scope, id, {
            ...props,
            handler:'handler',
            timeout:Duration.seconds(5),
            entry:`${__dirname}/../src/${id}.ts`
        })
    }
}