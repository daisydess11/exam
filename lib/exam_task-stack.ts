import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3Deployment from 'aws-cdk-lib/aws-s3-deployment';
import { Tags } from 'aws-cdk-lib';
import * as sns from 'aws-cdk-lib/aws-sns';
import { LambdaIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { HttpMethod } from 'aws-cdk-lib/aws-apigatewayv2';
import { Table, AttributeType, BillingMode } from 'aws-cdk-lib/aws-dynamodb';
import { BaseFunction } from './BaseFunction';
import { ServicePrincipal } from 'aws-cdk-lib/aws-iam';


//const secondaryIndex = 'timestamp';
const emailReceiver = "desislava.koshnicharova@gmail.com"

export class ExamTaskStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    
   

    // sns topic and subscription for valid json

    const validRecordTopic = new sns.Topic(this, 'validRecordTopic');

    const validRecordSubscription = new sns.Subscription(this,'validRecordSubscription',{
      topic:validRecordTopic,
      protocol: sns.SubscriptionProtocol.EMAIL,
      endpoint:emailReceiver
    })

    Tags.of(validRecordSubscription).add('lecture', 'exam');
    Tags.of(validRecordTopic).add('lecture','exam');
     // sns topic and subscription for deleted json

    const deletedRecordTopic = new sns.Topic(this, 'deletedRecordTopic');

    const deletedRecordSubscription = new sns.Subscription(this,'deletedRecordSubscription',{
      topic:deletedRecordTopic,
      protocol: sns.SubscriptionProtocol.EMAIL,
      endpoint:emailReceiver
    })

    Tags.of(deletedRecordSubscription).add('lecture', 'exam');
    Tags.of(deletedRecordTopic).add('lecture', 'exam');
    
    // dynamodb table for invalid records
    const invalidRecordsTable = new Table(this, 'invalidRecords', {
      partitionKey:{
        name:'PK',
        type:AttributeType.STRING
      },
      sortKey:{
        name:'SK',
        type:AttributeType.STRING
      },
      tableName: 'invalidRecords',
      billingMode: BillingMode.PAY_PER_REQUEST
    })

    // invalidRecordsTable.addGlobalSecondaryIndex({
    //   partitionKey:{
    //     name:secondaryIndex,
    //     type:AttributeType.STRING
    //   },
    //   indexName:secondaryIndex
    // });
    
    Tags.of(invalidRecordsTable).add('lecture', 'exam');


    

    //delete from table lambda 

     const deleteRecordFunction = new BaseFunction(this, 'deleteRecordFunction',{
       TABLE_NAME: invalidRecordsTable.tableName,
       TOPIC_ARN: deletedRecordTopic.topicArn
      });
 

    invalidRecordsTable.grantReadData(deleteRecordFunction)
    invalidRecordsTable.grantWriteData(deleteRecordFunction)

     //Lambda functions process record
    const processRecordFunction = new BaseFunction(this, 'processRecordFunction',{
        TABLE_NAME:invalidRecordsTable.tableName,
        TOPIC_ARN: validRecordTopic.topicArn,
        AWS_ACCOUNT: this.account,
        SCHEDULED_FUNC_ARN:deleteRecordFunction.functionArn
      });
      

    invalidRecordsTable.grantWriteData(processRecordFunction);
    validRecordTopic.grantPublish(processRecordFunction);  

    processRecordFunction.addPermission('AllowEventBridgeRuleInvoke', {
      principal: new ServicePrincipal('events.amazon.com'),
      action: 'lambda:InvokeFunction',
      sourceArn:`arn:aws:events:${this.region}:${this.account}:rule/*`
    })


    //api implementation
    const processRecordsApi = new RestApi(this,'ProcessRecordsApi',{
      restApiName:'ProcessRecords'
    })

    const orderResource = processRecordsApi.root.addResource('processRecord');
    orderResource.addMethod(HttpMethod.POST, new LambdaIntegration(processRecordFunction, {
      proxy:true
    }));
  }
  
}
