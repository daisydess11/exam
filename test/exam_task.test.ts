 import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as ExamTask from '../lib/exam_task-stack';


test('Template match snapshot', () => {
     const app = new cdk.App();
   const stack = new ExamTask.ExamTaskStack(app, 'MyTestStack');
  const template = Template.fromStack(stack);

 expect(template).toMatchSnapshot()
});
