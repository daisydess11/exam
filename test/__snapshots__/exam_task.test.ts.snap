// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template match snapshot 1`] = `
{
  "Outputs": {
    "ProcessRecordsApiEndpointB0CDEF4A": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "ProcessRecordsApi3AC528CD",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "ProcessRecordsApiDeploymentStageprod477B0C12",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ProcessRecordsApi3AC528CD": {
      "Properties": {
        "Name": "ProcessRecords",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ProcessRecordsApiAccountC9EF6185": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "ProcessRecordsApi3AC528CD",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "ProcessRecordsApiCloudWatchRole905DCA50",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "ProcessRecordsApiCloudWatchRole905DCA50": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "ProcessRecordsApiDeploymentA35A3E4B52068ad94b32aa67141202afcd53f58d": {
      "DependsOn": [
        "ProcessRecordsApiprocessRecordPOST80AA3158",
        "ProcessRecordsApiprocessRecord2BC9C662",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "ProcessRecordsApi3AC528CD",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ProcessRecordsApiDeploymentStageprod477B0C12": {
      "DependsOn": [
        "ProcessRecordsApiAccountC9EF6185",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "ProcessRecordsApiDeploymentA35A3E4B52068ad94b32aa67141202afcd53f58d",
        },
        "RestApiId": {
          "Ref": "ProcessRecordsApi3AC528CD",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ProcessRecordsApiprocessRecord2BC9C662": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "ProcessRecordsApi3AC528CD",
            "RootResourceId",
          ],
        },
        "PathPart": "processRecord",
        "RestApiId": {
          "Ref": "ProcessRecordsApi3AC528CD",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ProcessRecordsApiprocessRecordPOST80AA3158": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "processRecordFunction32EFC06C",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "ProcessRecordsApiprocessRecord2BC9C662",
        },
        "RestApiId": {
          "Ref": "ProcessRecordsApi3AC528CD",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ProcessRecordsApiprocessRecordPOSTApiPermissionMyTestStackProcessRecordsApi6E172954POSTprocessRecord3AECD24A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "processRecordFunction32EFC06C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ProcessRecordsApi3AC528CD",
              },
              "/",
              {
                "Ref": "ProcessRecordsApiDeploymentStageprod477B0C12",
              },
              "/POST/processRecord",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ProcessRecordsApiprocessRecordPOSTApiPermissionTestMyTestStackProcessRecordsApi6E172954POSTprocessRecord407F3364": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "processRecordFunction32EFC06C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ProcessRecordsApi3AC528CD",
              },
              "/test-invoke-stage/POST/processRecord",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "deleteRecordFunctionAB8179E5": {
      "DependsOn": [
        "deleteRecordFunctionServiceRoleDefaultPolicy24DEEF47",
        "deleteRecordFunctionServiceRoleEA51C682",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b9044e39602bbae1c0a9c54f40458adcb197f2b2ddd0ffd5d2598b238c270acc.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE_NAME": {
              "Ref": "invalidRecordsC035AFDC",
            },
            "TOPIC_ARN": {
              "Ref": "deletedRecordTopic393C1624",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "deleteRecordFunctionServiceRoleEA51C682",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "deleteRecordFunctionServiceRoleDefaultPolicy24DEEF47": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "invalidRecordsC035AFDC",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "invalidRecordsC035AFDC",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "deleteRecordFunctionServiceRoleDefaultPolicy24DEEF47",
        "Roles": [
          {
            "Ref": "deleteRecordFunctionServiceRoleEA51C682",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "deleteRecordFunctionServiceRoleEA51C682": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "deletedRecordSubscription410F33EC": {
      "Properties": {
        "Endpoint": "desislava.koshnicharova@gmail.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "deletedRecordTopic393C1624",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "deletedRecordTopic393C1624": {
      "Properties": {
        "Tags": [
          {
            "Key": "lecture",
            "Value": "exam",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
    "invalidRecordsC035AFDC": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "TableName": "invalidRecords",
        "Tags": [
          {
            "Key": "lecture",
            "Value": "exam",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "processRecordFunction32EFC06C": {
      "DependsOn": [
        "processRecordFunctionServiceRoleDefaultPolicy227C1F8A",
        "processRecordFunctionServiceRoleF5184E58",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "47449e525ebe535cfe19f6a657de6cb3a4abac18b7fc2816dafdada8e7fd3e56.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_ACCOUNT": {
              "Ref": "AWS::AccountId",
            },
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SCHEDULED_FUNC_ARN": {
              "Fn::GetAtt": [
                "deleteRecordFunctionAB8179E5",
                "Arn",
              ],
            },
            "TABLE_NAME": {
              "Ref": "invalidRecordsC035AFDC",
            },
            "TOPIC_ARN": {
              "Ref": "validRecordTopicDDC92C9C",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "processRecordFunctionServiceRoleF5184E58",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "processRecordFunctionServiceRoleDefaultPolicy227C1F8A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "invalidRecordsC035AFDC",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "validRecordTopicDDC92C9C",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "processRecordFunctionServiceRoleDefaultPolicy227C1F8A",
        "Roles": [
          {
            "Ref": "processRecordFunctionServiceRoleF5184E58",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "processRecordFunctionServiceRoleF5184E58": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "validRecordSubscription84DAFF35": {
      "Properties": {
        "Endpoint": "desislava.koshnicharova@gmail.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "validRecordTopicDDC92C9C",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "validRecordTopicDDC92C9C": {
      "Properties": {
        "Tags": [
          {
            "Key": "lecture",
            "Value": "exam",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
